demo/app.py

```diff
@@ -35,8 +67,52 @@ def index():
         response['code'] = res.status_code
         response['data'] = json.loads(res.text)['data']
 
-    return render_template('index.html', form=onid_form, onid=onid, res=response)
+    return render_template('index.html', form=onid_form, onid=onid, res=response, logout_url=logout_url)
+
+@app.route('/logout', methods=['GET'])
+def logout():
+    if 'user' in session:
+        del session['user']
+    return flask.redirect(app.config['CAS_URL'] + "/logout")
+
+class CASError(Exception):
+    def __init__(self, msg, code=''):
+        self.msg = msg
+        self.code = code
+    def __str__(self):
+        if self.code and self.msg:
+            return str(self.msg) + " (" + str(self.code) + ")"
+        if self.msg:
+            return str(self.msg)
+        return str(self.code)
+
+def validate_cas(cas_url, ticket, service):
+    r = requests.get(cas_url+'/serviceValidate', params={'ticket': ticket, 'service': service})
+    if r.status_code != 200:
+        raise CASError('invalid response')
+
+    # note: processing xml from untrusted sources is unsafe
+    # https://docs.python.org/2/library/xml.html#xml-vulnerabilities
+    try:
+        root = elementtree.fromstring(r.text)
+    except elementtree.ParseError:
+        raise CASError('invalid response')
+
+    ns = '{http://www.yale.edu/tp/cas}'
+    if root.tag != ns+'serviceResponse':
+        raise CASError('invalid response')
+
+    authenticationFailure = root.find(ns+'authenticationFailure')
+    authenticationSuccess = root.find(ns+'authenticationSuccess')
+
+    if authenticationFailure is not None:
```

Comment:
`if authenticationFailure:`?

Reply:
Most of your comments are on code I _borrowed_ :) from the [oauth2 consent app](https://github.com/osu-mist/oauth2-consent-prototype) that @magical wrote. Either way, I was interested in your suggestion so I found this that was helpful: https://stackoverflow.com/a/7816439/6762736

Also, [this](https://www.python.org/dev/peps/pep-0008/#programming-recommendations):
>Also, beware of writing if x when you really mean if x is not None -- e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

Reply:
Awesome. I was not 100% sure the timing of using this kind of shorthand. These articles explain the difference well and also cleared up some doubt of mine. Learn some new stuff, thanks for pointing this out.

Reply:
Fun fact - if you actually try this you'll get a warning:
```
>>> from xml.etree import ElementTree
>>> a = ElementTree.fromstring('<a><b/></a>')
>>> b = a.find('b')
>>> if b: pass
... 
__main__:1: FutureWarning: The behavior of this method will change in future versions.  Use specific 'len(elem)' or 'elem is not None' test instead.
>>>
```